### PR TITLE
Lior/raptor 3656

### PR DESCRIPTION
**Pull request by liororama**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
Adding r2d2 a Swiss Army Knife inference model - performing actions such as memory consumption, exception throwing and more.

## Rationale

r2d2 will be used to test the drum infrastructure and cm in general for errors and extreme scenarios.

### CODE REVIEW

**Comment by liororama** [comment]
jarvis run functional_drop_in

**Comment by liororama** [comment]
jarvis test functional_drop_in

File: tools/r2d2/README.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# R2D2: A testing inference model
+
+r2d2 inference model provides a quick way to test your environment.
+r2d2 is not really a model, it is a testing program which interpret 
+the prediction requests as actions to execute inside the drum environment.
+
+r2d2 expects is input dataframe to contain the 2 columns: "cmd" and "arg".
```
Comment (yakov-g) [inline tools/r2d2/README.md:7]: Grammar.
I think `is` is not needed.

File: tools/r2d2/README.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    memory,3000
+
+
+## What kind of actions are supported:
+r2d2 supports the following actions:
+* Consume memory
+* Throw exceptions
+* Sleep for a given amount of time, potentially causing timeout.
+
+###Use memory
```
Comment (yakov-g) [inline tools/r2d2/README.md:41]: Space is needed after ### otherwise it is not rendered

File: tools/r2d2/custom.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
+import pandas as pd
+import sys
+import time
+
+prediction_value = None
+
+
+class R2D2:
+    MEGA = 10 ** 6
```
Comment (yakov-g) [inline tools/r2d2/custom.py:9]: Just for consistency - let's make it 1048576

File: tools/r2d2/custom.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    MEGA_STR = " " * MEGA
+
+    CMD_COL = "cmd"
+    ARG_COL = "arg"
+
+    def __init__(self):
+        self._mem_array = []
+        print("R2D2 __init__")
```
Comment (yakov-g) [inline tools/r2d2/custom.py:17]: Remove this debug message?

File: tools/r2d2/custom.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        return len(self._mem_array)
+
+    def _clear_memory(self):
+        self._mem_array = []
+
+    def _alloc_additional_memory(self, additional_mb):
+        for i in range(0, additional_mb):
+            try:
+                self._mem_array.append(R2D2.MEGA_STR + str(i))
+                print("Adding: {}".format(i))
```
Comment (yakov-g) [inline tools/r2d2/custom.py:29]: Remove this, otherwise it is printed thousands of times.

File: tools/r2d2/custom.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            except MemoryError:
+                print("No more memory here..")
```
Comment (yakov-g) [inline tools/r2d2/custom.py:31]: More informative message. You could write: "Can not allocate more memory. Total allocated memory: XXX "

File: tools/r2d2/custom.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+                break
+
+            i += 1
+
+    def alloc_memory(self, arg):
+        # Always taking the info to run from the first command
+        memory_mb = int(arg)
+        print("Found memory_mb in data: {}".format(memory_mb))
+        if memory_mb <= 0:
```
Comment (yakov-g) [inline tools/r2d2/custom.py:40]: Add into README, that sending  negative/zero number - clears mem.

File: tools/r2d2/custom.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+            print("Clearing memory")
+            self._clear_memory()
+        elif memory_mb > 0:
+            additional_mem = memory_mb - self._mem_size_mb()
```
Comment (yakov-g) [inline tools/r2d2/custom.py:44]: So sending 2 times command:
memory,2000 allocates just 2000 and not 4000
Mention it in the readme

File: tools/r2d2/custom.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    r2d2.handle_prediction_request(data)
+
+    prediction_value += 1
+    predictions = pd.DataFrame(
+        [prediction_value for _ in range(data.shape[0])], columns=["Predictions"]
+    )
+    return predictions
+
+
+def post_process(predictions, model):
```
Comment (yakov-g) [inline tools/r2d2/custom.py:125]: This hook can be removed

File: tools/r2d2/custom.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    return pd.read_csv(input_file)
+
+
+def load_model(input_dir):
+    global prediction_value
+    prediction_value += 1
+    return "dummy"
+
+
+def transform(data, model):
```
Comment (yakov-g) [inline tools/r2d2/custom.py:107]: This hooke can be removed

File: tools/r2d2/custom.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    prediction_value = 1
+
+
+def read_input_data(input_file):
+    global prediction_value
+    prediction_value += 1
+    return pd.read_csv(input_file)
+
+
+def load_model(input_dir):
```
Comment (yakov-g) [inline tools/r2d2/custom.py:101]: This can just return "dummy"

File: tools/r2d2/custom.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+# Global instance
+r2d2 = R2D2()
+
+
+def init(**kwargs):
+    global prediction_value
+    prediction_value = 1
+
+
+def read_input_data(input_file):
```
Comment (yakov-g) [inline tools/r2d2/custom.py:95]: This hook can be removed

File: tools/r2d2/custom.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+        else:
+            print("Cmd: {} is not supported".format(cmd))
+            raise Exception("Bad CMD provided {}".format(cmd))
+
+
+# Global instance
+r2d2 = R2D2()
+
+
+def init(**kwargs):
```
Comment (yakov-g) [inline tools/r2d2/custom.py:90]: This hook can be removed

File: tools/r2d2/custom.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+    data_mem = pd.DataFrame({"cmd": ["memory"], "arg": [1000]}, columns=["cmd", "arg"])
+    data_exception = pd.DataFrame({"cmd": ["exception"], "arg": [5]}, columns=["cmd", "arg"])
+    data_timeout = pd.DataFrame({"cmd": ["timeout"], "arg": [10]}, columns=["cmd", "arg"])
+
+    print(data_mem)
+    print(data_exception)
+    print(data_timeout)
+
+    init()
```
Comment (yakov-g) [inline tools/r2d2/custom.py:142]: This hook can be removed

File: tools/r2d2/custom.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
+import pandas as pd
+import sys
+import time
+
+prediction_value = None
```
Comment (yakov-g) [inline tools/r2d2/custom.py:5]: This hook can be removed

**Review by yakov-g — CHANGES_REQUESTED** [review]

File: tools/r2d2/custom.py (hunk 5 above)
Comment (liororama) [reply tools/r2d2/custom.py:29]: This was exactly the point for printing it.. r2d2 is a debug tool.. you want to see what was the last memory allocation the went ok.. and to see the rate of memory allocations.

File: tools/r2d2/custom.py (hunk 6 above)
Comment (liororama) [reply tools/r2d2/custom.py:31]: sure.

File: tools/r2d2/custom.py (hunk 5 above)
Comment (yakov-g) [reply tools/r2d2/custom.py:29]: But last memory allocation you can print in the MemoryError exception. Or in the end of the loop.

And you will not see rate of memory allocations when do this in pred server mode. Especially, i think, OS is smart enough to allocate BIG chunk if there were many requests for smaller chunks. So I don't see these printouts help you.

Comment (liororama) [reply tools/r2d2/custom.py:29]: I will print every 100MB of memory.. 
We don't see the exception.. the OOM killer or docker is just killing the entire process..

**Comment by yakov-g** [comment]
jarvis test functional_drop_in

**Review by yakov-g — APPROVED** [review]

**Review by meltzg — APPROVED** [review]
any way to add tests?

**Comment by liororama** [comment]
I am intending to add a test when I am adding my next PR to this one.. since it will add a way to limit the memory of a container so I will have a single test to test both the --memory and r2d2.

**Comment by datarobot-jarvis** [comment]
:dollar: Total cost of executing tests in Jenkins: $**4.27** USD.

[What is this all about?](http://bit.ly/2BahK4V)

**Pull request merged**